src/Storages/ObjectStorage/DataLakes/Iceberg/DeleteFiles/PositionalDeleteTransform.h

```diff
@@ -0,0 +1,186 @@
+                auto filename_column = delete_chunk.getColumns()[filename_index];
+                auto last_filename = filename_column->getDataAt(delete_chunk.getNumRows() - 1).toString();
+                last_filename = cropPrefix(std::move(last_filename));
+                if (last_filename < source_filename)
+                    break;
+
+                auto last_position = position_column->get64(delete_chunk.getNumRows() - 1);
+                if (last_position < chunk_rows_read_before && last_filename == source_filename)
+                    break;
+
+                auto first_filename = filename_column->getDataAt(0).toString();
+                first_filename = cropPrefix(std::move(first_filename));
+                if (first_filename > source_filename)
+                {
+                    unprocessed_delete_chunk[delete_source_id] = std::move(delete_chunk);
+                    break;
+                }
+
+                auto first_position = position_column->get64(0);
+                if (first_position >= chunk_rows_read_before + chunk.getNumRows() && last_filename == source_filename)
+                {
+                    unprocessed_delete_chunk[delete_source_id] = std::move(delete_chunk);
+                    break;
+                }
+
+                bool should_cache = false;
+                for (size_t i = 0; i < delete_chunk.getNumRows(); ++i)
+                {
+                    auto position_to_delete = position_column->get64(i);
+                    auto filename_to_delete = filename_column->getDataAt(i).toString();
+                    if (cropPrefix(std::move(filename_to_delete)) == source_filename)
+                    {
+                        if (position_to_delete - chunk_rows_read_before < chunk.getNumRows())
+                        {
+                            should_delete[position_to_delete - chunk_rows_read_before] = false;
+                            --num_rows_after_filtration;
+                        }
+                        else
+                        {
+                            should_cache = true;
+                        }
+                    }
+                }
+                if (should_cache)
+                {
+                    unprocessed_delete_chunk[delete_source_id] = std::move(delete_chunk);
+                    break;
+                }
+            }
+        }
+
+        auto columns = chunk.detachColumns();
+        for (auto & column : columns)
+            column = column->filter(should_delete, -1);
+
+        chunk.setColumns(std::move(columns), num_rows_after_filtration);
+    }
+
+private:
+    std::vector<std::shared_ptr<IInputFormat>> delete_file_sources;
+    String description;
+    Block header;
+
+    std::vector<std::optional<Chunk>> unprocessed_delete_chunk;
+
+    std::vector<std::optional<int>> filename_column_index;
+    std::vector<std::optional<int>> position_column_index;
```

Comment:
If there are too many positions, isn't this implementation too good? using roaring bitmap?

Reply:
size of this vectors is not equal to size of delete files, it is equal to the number of delete files that should be applied to the date file